src/java/org/apache/cassandra/index/sai/StorageAttachedIndex.java

```diff
@@ -0,0 +1,477 @@
+import org.apache.cassandra.schema.TableMetadata;
+import org.apache.cassandra.utils.Pair;
+import org.apache.cassandra.utils.concurrent.OpOrder;
+
+public class StorageAttachedIndex implements Index
+{
+    private static final Logger logger = LoggerFactory.getLogger(StorageAttachedIndex.class);
+
+    private static class StorageAttachedIndexBuildingSupport implements IndexBuildingSupport
+    {
+        @Override
+        public SecondaryIndexBuilder getIndexBuildTask(ColumnFamilyStore cfs,
+                                                       Set<Index> indexes,
+                                                       Collection<SSTableReader> sstablesToRebuild,
+                                                       boolean isFullRebuild)
+        {
+            // Return an empty builder for in-memory implementation
+            return new SecondaryIndexBuilder()
+            {
+                @Override
+                public void build()
+                {
+                }
+
+                @Override
+                public CompactionInfo getCompactionInfo()
+                {
+                    return null;
+                }
+            };
+        }
+    }
+
+    // Used to build indexes on newly added SSTables:
+    private static final StorageAttachedIndexBuildingSupport INDEX_BUILDER_SUPPORT = new StorageAttachedIndexBuildingSupport();
+
+    private static final Set<String> VALID_OPTIONS = ImmutableSet.of(IndexTarget.TARGET_OPTION_NAME,
+                                                                     IndexTarget.CUSTOM_INDEX_OPTION_NAME);
+
+    public static final Set<CQL3Type> SUPPORTED_TYPES = ImmutableSet.of(CQL3Type.Native.ASCII, CQL3Type.Native.BIGINT, CQL3Type.Native.DATE,
+                                                                        CQL3Type.Native.DOUBLE, CQL3Type.Native.FLOAT, CQL3Type.Native.INT,
+                                                                        CQL3Type.Native.SMALLINT, CQL3Type.Native.TEXT, CQL3Type.Native.TIME,
+                                                                        CQL3Type.Native.TIMESTAMP, CQL3Type.Native.TIMEUUID, CQL3Type.Native.TINYINT,
+                                                                        CQL3Type.Native.UUID, CQL3Type.Native.VARCHAR, CQL3Type.Native.INET,
+                                                                        CQL3Type.Native.VARINT, CQL3Type.Native.DECIMAL, CQL3Type.Native.BOOLEAN);
+
+    private static final Set<Class<? extends IPartitioner>> ILLEGAL_PARTITIONERS =
+            ImmutableSet.of(OrderPreservingPartitioner.class, LocalPartitioner.class, ByteOrderedPartitioner.class, RandomPartitioner.class);
+
+    private final ColumnFamilyStore baseCfs;
+    private final IndexMetadata config;
+    private final IndexContext indexContext;
+
+    public StorageAttachedIndex(ColumnFamilyStore baseCfs, IndexMetadata config)
+    {
+        this.baseCfs = baseCfs;
+        this.config = config;
```

Comment:
nit: Might be possible to just remove the `IndexMetadata` field and delegate to `IndexContext`